Apps/Sandcastle/gallery/Drawing on Terrain.html

```diff
@@ -0,0 +1,156 @@
+function startup(Cesium) {
+    'use strict';
+//Sandcastle_Begin
+var viewer = new Cesium.Viewer('cesiumContainer',  {
+    selectionIndicator : false,
+    infoBox : false,
+    terrainProvider : Cesium.createWorldTerrain()
+});
+viewer.cesiumWidget.screenSpaceEventHandler.removeInputAction(Cesium.ScreenSpaceEventType.LEFT_DOUBLE_CLICK);
+function createPoint(worldPosition) {
+    var point = viewer.entities.add({
+        position : worldPosition,
+        point : {
+            color : Cesium.Color.WHITE,
+            pixelSize : 5,
+            heightReference: Cesium.HeightReference.CLAMP_TO_GROUND
+        }
+    });
+    return point;
+}
+var drawingMode = 'line';
+function drawShape(positionData) {
+    var shape;
+    if (drawingMode === 'line') {
+        shape = viewer.entities.add({
+            polyline : {
+                positions : positionData,
+                clampToGround : true,
+                width : 3
+            }
+        });
+    }
+    else if (drawingMode === 'polygon') {
+        shape = viewer.entities.add({
+            polygon: {
+                hierarchy: positionData,
+                outline: true,
+                material: new Cesium.ColorMaterialProperty(Cesium.Color.WHITE.withAlpha(0.7))
+            }
+        });
+    }
+    return shape;
+}
+var activeShapePoints = [];
+var activeShape;
+var floatingPoint;
+var handler = new Cesium.ScreenSpaceEventHandler(viewer.canvas);
+handler.setInputAction(function(event) {
+    // We use `viewer.scene.pickPosition` here instead of `viewer.camera.pickEllipsoid` so that
+    // we get the correct point when mousing over terrain.
+    var earthPosition = viewer.scene.pickPosition(event.position);
+    // `earthPosition` will be undefined if our mouse is not over the globe.
+    if (Cesium.defined(earthPosition)) {
+        if (activeShapePoints.length === 0) {
+            floatingPoint = createPoint(earthPosition);
+            activeShapePoints.push(earthPosition);
+            var dynamicPositions = new Cesium.CallbackProperty(function () {
+                return activeShapePoints;
+            }, false);
+            activeShape = drawShape(dynamicPositions);
+        }
+        activeShapePoints.push(earthPosition);
+        createPoint(earthPosition);
+    }
+}, Cesium.ScreenSpaceEventType.LEFT_CLICK);
+
+handler.setInputAction(function(event) {
+    if (Cesium.defined(floatingPoint)) {
+        var newPosition = viewer.scene.pickPosition(event.endPosition);
+        if (Cesium.defined(newPosition)) {
+            floatingPoint.position.setValue(newPosition);
+            activeShapePoints.pop();
+            activeShapePoints.push(newPosition);
+        }
+    }
+}, Cesium.ScreenSpaceEventType.MOUSE_MOVE);
+// Redraw the shape so it's not dynamic and remove the dynamic shape.
+function terminateShape() {
+    activeShapePoints.pop();
+    drawShape(activeShapePoints);
+    viewer.entities.remove(floatingPoint);
+    viewer.entities.remove(activeShape);
```

Comment:
The shapes flash when you terminate the shape. Is it possible to re-use the same entity and create the position property, or do we have to delete the old one and create a new dynamic one?

Reply:
I just tested that out, and it'll still take a few seconds to render when you switch the `CallbackProperty` with just an array of points.

I talked briefly with @mramato about this. We _could_ do any of the following to remove that flicker:

* Keep all the shapes with `isConstant : false`
* Don't remove the dynamic shape right away, keep it there for a few seconds to hide the gap until the new one renders. 

But none of those seem like good practices we want to encourage in an official Sandcastle example.

It might be nice to have a "flush" command that would try to immediately render an entity.

Reply:
This is a long standing issue in "sometimes changing" geometry data in Cesium.  I recommend this PR keeps whatever is cleanest for example code.  I would love to see us improve interactive geometry like this, but it's probably going to result in a request-render style method for entities that needs to be thought through.